aip/0194.md

```diff
@@ -0,0 +1,88 @@
+**should** have user-level retry logic that retries the entire transaction
+block from the start.
+
+Clients **should not** retry requests in which repeated runs would cause
+unintended state changes.
+
+This document does not cover client streaming and bidirectional streaming
+requests, as the retryability of these requests are highly dependent on the
+type of data being streamed.
+
+### Retryable codes
+
+These codes **should** be retried for retryable requests:
+
+- `INTERNAL`: This generally means that some internal part of the system has
+failed. These are retryable under the expectation that the system will recover
+from failures.
+- `UNAVAILABLE`: These codes generally result from network hiccups, and are
+generally transient. They are retryable under the expectation that the
+connection will become available (soon).
+
+## Appendix
+
+### Non-retryable codes
+
+These codes **should not** be retried for any request:
+
+- `OK`: The request succeeded.
+- `CANCELLED`: A user can cancel a request, which **must** be honoured.
+- `DEADLINE_EXCEEDED`: A user can set a deadline, which **must** be honoured.
+- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never
+  succeed.
+- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be
+  surfaced to the user.
+
+### Generally non-retryable codes:
+
+Some codes **may** be retried if a system is designed without synchronization
+between various components. For example, a system in which a read is designed to
+hang forever until a resource is created. However, these are not the types of
+systems generally encouraged, and a client **should not** retry these types of
+errors.
+
+These codes generally **should not** be retried for any request:
+
+- `RESOURCE_EXHAUSTED`: This may be a signal that the user's quota is exhausted.
+  Retries therefore may not be expected to work for several hours; meanwhile the
+  retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for
+  other reasons than quota and the expected time for the resource to become
+  available is much shorter, it may be retryable.
+- `NOT_FOUND`: The resource may be created in the future.
+- `ALREADY_EXISTS`: The resource may be deleted in the future.
+- `PERMISSION_DENIED`: The user may be authenticated in the future.
+- `FAILED_PRECONDITION`: The system state may change in the future.
+- `OUT_OF_RANGE`: The range may be extended in the future.
+- `UNIMPLEMENTED`: The RPC may be implemented in the future.
+- `UNKNOWN`: `UNKNOWN` and `INTERNAL` are both generic error codes - they are
+  errors that may be generated from a wide breadth of causes. Known services
+  within the system **should** generate `INTERNAL` errors when the error doesn't
+  fit any other code. `UNKNOWN` errors are reserved for truly
+  unknown-to-the-system errors, and therefore **should** immediately surfaced to
+  the user.
+
+## Further reading
+
+Also see: [Request Validation](https://aip.dev/154).
```

Comment:
This suggestion is just for following the style of other "further reading sections".

```suggestion
- For parallel or retried request disambiguation, see [AIP-154][].

[aip-154]: ./0154.md
```

Reply:
Done